Yoast/wordpress-seo
# CoAuthors Plus integration

## Context


* Allow support for multiple authors in Schema.

## Summary


This PR can be summarized in the following changelog entry:

* Adds an integration for [CoAuthors Plus](https://wordpress.org/plugins/co-authors-plus/). When enabled, multiple authors will be added to the schema output.
* Changes the `@id` for images for `Person` and `Organization` including the `logo` attribute for the latter, to have a more proper format.
* Adds a `wpseo_schema_graph` filter to allow filtering the entire Schema graph.

## Relevant technical choices:

* Decided to change the #personlogo schema ID to be more specific to the image, as it would clash otherwise. Made the same change for `logo` attributes. When CoAuthors is active, we suffix the ID with a hash of the image url.

## Test instructions

### Test instructions for the acceptance test before the PR gets merged
This PR can be acceptance tested by following these steps:

_Notes_
Use the following tools to validate the schemas:
* Use https://validator.schema.org/ to check that there are no syntax errors.
* Use https://classyschema.org/Visualisation check that the graph is cohesive and complete.
* Use https://search.google.com/test/rich-results to check any Google-specific requirements/errors.

**Test 1**
* Install & activate [CoAuthors Plus](https://wordpress.org/plugins/co-authors-plus/).
* Assign multiple authors to a post.
* See that the post's schema output now has multiple `Person` blocks and the `Article` schema references them all by ID.
* See that this schema validates.

**Test 2**
* Also check that [this comment](https://github.com/Yoast/wordpress-seo/pull/18161#issuecomment-1067798649) has been addressed. 

**Test 3**
* Also test this case:
  * set 2 authors for a post
  * set the site as representing a Person and choose the second author as the Person in question
  * check the Schema:
    * the Article piece should have 2 authors referenced by ID
    * one of them should point to the `[Person, Organization]` piece
    * there shouldn't be another `Person` piece referencing the same user
  * change the site representation so it's set to Organization
  * check the Schema:
      * the Article piece should have 2 authors referenced by ID
      * there should be 2 `Person` pieces, one for each of the authors
      * _Note: this test covers the case where one of users has their ID still set in the DB for Person even if the site is representing an Organization_

**Test 4**
* Create a post with featured Image, but also with Open Graph or Twitter image
With Trunk:
* The `thumbnailUrl` attribute of the `Article` object IS NOT THE SAME  with the `ImageObject` that has the `PrimaryImage` id
With this PR:
* The `thumbnailUrl` attribute of the `Article` object IS THE SAME with the `ImageObject` that has the `PrimaryImage` id


**Specific subcases to test (validate the schemas for CoAuthors Plus both active and inactive)**

1. Post has 1 author which is the Site Represents user
2. Post has 1 author which is NOT the Site Represents user
3. Post has 2 authors NONE of which is the Site Represents user (only with CoAuthors Plus active)
4. Post has 2 authors THE FIRST of which is the Site Represents user (only with CoAuthors Plus active)
5. Post has 2 authors THE SECOND of which is the Site Represents user (only with CoAuthors Plus active)
6. Post with 2 authors that is a Tech Article (with CoAuthors Plus active, it should show two Authors/Persons in the schema)


### Test instructions for QA when the code is in the RC



## Impact check

This PR affects the following parts of the plugin, which may require extra testing:

Validate the schemas for the below (for CoAuthors Plus both active and inactive):

1. Post while site is organization
2. Post with HowTo block
3. Post with FAQ block
4. Events Calendar schema (Event’s schema) 
5. Events Calendar schema (Month’s schema)
6. Post with no featured image but with resized image in content
7. Post with Local SEO enabled with single location
8. Post with Local SEO enabled with multiple locations
9. Location CPT with Local SEO enabled with multiple locations, same branch
10. Location CPT with Local SEO enabled with multiple locations, different branch
11. Product with WooCommerce SEO enabled
12. Post with video with Video SEO enabled
13. Author's pages, eg. http://basic.wordpress.test/author/admin/
14. Categories


## UI changes


## Other environments


## Documentation


## Quality assurance


Fixes #
